File: src/runtime2/cs/pytra-core/std/pathlib.cs
using System;
using System.IO;

namespace Pytra.CsModule
{
    // Python pathlib の最小互換。
    public class py_path
    {
        private readonly string _value;

        public py_path(string value)
        {
            _value = value ?? string.Empty;
        }

        public py_path resolve()
        {
            return new py_path(Path.GetFullPath(_value));
        }

        public py_path parent()
        {
            string dir = Path.GetDirectoryName(_value);
            return new py_path(dir ?? string.Empty);
        }

        public string name()
        {
            return Path.GetFileName(_value);
        }

        public string stem()
        {
            return Path.GetFileNameWithoutExtension(_value);
        }

        public bool exists()
        {
            return File.Exists(_value) || Directory.Exists(_value);
        }

        public string read_text(string encoding = "utf-8")
        {
            return File.ReadAllText(_value);
        }

        public void write_text(string content, string encoding = "utf-8")
        {
            File.WriteAllText(_value, content ?? string.Empty);
        }

        public void mkdir(bool parents_flag = false, bool exist_ok = false)
        {
            try
            {
                if (parents_flag)
                {
                    Directory.CreateDirectory(_value);
                    return;
                }
                Directory.CreateDirectory(_value);
            }
            catch
            {
                if (!exist_ok)
                {
                    throw;
                }
            }
        }

        public static py_path operator /(py_path lhs, string rhs)
        {
            return new py_path(Path.Combine(lhs._value, rhs));
        }

        public override string ToString()
        {
            return _value;
        }
    }
}

File: src/runtime2/js/pytra-core/std/pathlib.js
// Python pathlib の最小互換実装（Node.js）。

const fs = require("fs");
const path = require("path");

class Path {
  constructor(value) {
    this.value = String(value ?? "");
  }

  resolve() {
    return new Path(path.resolve(this.value));
  }

  parent() {
    return new Path(path.dirname(this.value));
  }

  name() {
    return path.basename(this.value);
  }

  stem() {
    return path.parse(this.value).name;
  }

  exists() {
    return fs.existsSync(this.value);
  }

  read_text(_encoding = "utf-8") {
    return fs.readFileSync(this.value, "utf8");
  }

  write_text(content, _encoding = "utf-8") {
    fs.writeFileSync(this.value, String(content), "utf8");
  }

  mkdir(parents = false, exist_ok = false) {
    if (parents) {
      fs.mkdirSync(this.value, { recursive: true });
      return;
    }
    try {
      fs.mkdirSync(this.value);
    } catch (err) {
      if (!(exist_ok && err && err.code === "EEXIST")) {
        throw err;
      }
    }
  }

  toString() {
    return this.value;
  }
}

function pathJoin(base, child) {
  return new Path(path.join(String(base), String(child)));
}

module.exports = { Path, pathJoin };
